File: PYTHON/Player.py

"""this file contains classes that have to do with the Player object"""

from __future__ import annotations
from typing import *
import discord
import random
import Character

import Adventure

class Player:

    """
    contains info and methods on the player, his history, and current data.
    """

    def __init__(self, author : discord.Member):

        self.author : discord.Member = author
        self.nickname : str = author.name
        self.characters : List[Character.Character] = []
        self.character : Character.Character = None
        self.adventure : Adventure.Adventure = None
        self.adventuresPlayed : Set[Adventure.Adventure] = set()

        self.party : Party = Party([self])

    def Reset(self):
        self.party = None
        self.adventure = None

        for c in self.characters:
            c.player = None


class Party:
    def __init__(self, players):
        self.players : List[Player] = players
        self.partyLeader : Player or None = None
        self.partyLeader = random.choice(self.players)

    def AddPlayer(self, player : Player):
        self.players.append(player)
        player.party = self

    def SetPartyLeader(self, player):
        if player in self.players:
            self.partyLeader = player
        else:
            raise Exception("party leader not in party")

    def Kick(self, player : Player):
        self.players.remove(player)
        player.party = Party([player])
        if self.partyLeader == player:
            self.partyLeader = self.players[0]

    def __getitem__(self, item) -> Player:

        if type(item) == discord.Member or type(item) == discord.User:

            for p in self.players:
                if p.author == item:
                    return p
            raise KeyError(f"player {item} not in party")

        elif type(item) == int:
            return self.players[item]

        else:
            raise TypeError(f"party object can't index {type(item)} object")

    def __len__(self):
        return len(self.players)

File: C#/Player.cs
using System;
using System.Collections.Generic;
using System.Text;
using DSharpPlus.Entities;
using DSharpPlus.CommandsNext;
using DSharpPlus.CommandsNext.Attributes;

namespace CSDNDISCORD
{
    class Player
    {
        public DiscordMember member;
        public string name;
        public string username;
        public List<Character> characters = new List<Character>();

        public Adventure adventure = null;
        public HashSet<Adventure> adventuresPlayed = new HashSet<Adventure>();

        public Party party = null;
        public Character character;

        public Player(DiscordMember member, string username = "")
        {
            this.member = member;
            this.name = member.DisplayName;
            this.username = username;
        }

        public async void LeaveParty(CommandContext ctx)
        {
            if (party == null) { await ctx.Channel.SendMessageAsync("You are not in a party."); return; }
            party.players.Remove(this);
            if (party.leader == this) { party.leader = party.players[0]; 
                await ctx.Channel.SendMessageAsync("Left the party."); }
            
        }
    }

    class Party
    {
        public List<Player> players { get; private set; } = new List<Player>();
        public Player leader;

        public Party(List<Player> players)
        {
            this.players = players;
            leader = players[0];
        }

        public void AddPlayer (Player player)
        {
            players.Add(player);
            player.party = this;
        }

        public async void Kick (CommandContext ctx, Player player)
        {
            if (!players.Contains(player))
            {
                await ctx.Channel.SendMessageAsync("Player not in party");
                return;
            }
            if (player == leader)
            {
                await ctx.Channel.SendMessageAsync("You can't kick the party leader.");
            } else
            {
                players.Remove(player);
                await ctx.Channel.SendMessageAsync("Kicked " + player.member.Mention + " from the party");
                await player.member.SendMessageAsync("You were kicked from " + leader.member.Mention + "'s party");
            }
        }
    }
}
